app/experimenter/nimbus-ui/src/hooks/useCommonForm.tsx

```diff
@@ -0,0 +1,143 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import React from "react";
+import { RegisterOptions, FieldError } from "react-hook-form";
+import Form from "react-bootstrap/Form";
+import { useForm } from "react-hook-form";
+
+// exported for testing
+export const overviewFieldNames = [
+  "name",
+  "hypothesis",
+  "application",
+  "publicDescription",
+] as const;
+
+export const audienceFieldNames = [
+  "channel",
+  "firefoxMinVersion",
+  "targetingConfigSlug",
+  "populationPercent",
+  "totalEnrolledClients",
+  "proposedEnrollment",
+  "proposedDuration",
+] as const;
+
+export const metricsFieldNames = [
+  "primaryProbeSetIds",
+  "secondaryProbeSetIds",
+] as const;
+
+type OverviewFieldNames = typeof overviewFieldNames[number];
+type AudienceFieldNames = typeof audienceFieldNames[number];
+type MetricsFieldNames = typeof metricsFieldNames[number];
+
+type FieldNames = OverviewFieldNames | AudienceFieldNames | MetricsFieldNames;
+
+// TODO: 'any' type on `onChange={(selectedOptions) => ...`,
+// it wants this, but can't seem to coerce it into SelectOption type
+// type SelectedOption = {
+//   value: ValueType<any, true>;
+//   action: ActionMeta<any>;
+// };
+export type SelectOption = { label: string; value: string };
+
+export function useCommonForm(
+  defaultValues: Record<string, any>,
+  isServerValid: boolean,
+  submitErrors: Record<string, string[]>,
+  setSubmitErrors: React.Dispatch<React.SetStateAction<Record<string, any>>>,
+) {
+  const {
+    handleSubmit,
+    register,
+    reset,
+    errors,
+    formState: { isSubmitted, isDirty, touched, isValid: isClientValid },
+  } = useForm({
+    mode: "onTouched",
+    defaultValues,
+  });
+
+  const isValid = isServerValid && isClientValid;
+  const isDirtyUnsaved = isDirty && !(isValid && isSubmitted);
+
+  const hideSubmitError = <K extends FieldNames>(name: K) => {
+    if (submitErrors[name]) {
+      const modifiedSubmitErrors = { ...submitErrors };
+      delete modifiedSubmitErrors[name];
+      setSubmitErrors(modifiedSubmitErrors);
+    }
+  };
+
+  // Fields are required by default. Pass an empty object `{}` as `registerOptions`
```

Comment:
Not a blocker but after digging through this and writing some docs, it seems like almost almost no fields should actually be required (the exceptions are `name` for experiment and branch) so it might make sense to reverse this default

Reply:
@k88hudson Yeah, we brought this up in Slack a day or two ago I think, but we didn't want to tack one more thing onto this since I was already almost done and it'd need more tests etc. I think we already have a ticket for it, if not I'll make another. 👍